### PR TITLE
Improvement/peer blacklist

### PR DESCRIPTION
**Pull request by denis-trapeznikov**
This commit adds a temporary blacklist for alien peers.

### CODE REVIEW

File: libs/core/tests/unit/temporary_blacklist_tests.cpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+using namespace std::chrono_literals;
+
+using TestedClass = fetch::core::TemporaryBlacklist<int>;
+
+TEST(TemporaryBlacklistTests, Main)
+{
+  TestedClass blacklist(1s);
+
+  blacklist.Blacklist(10);
+  std::this_thread::sleep_for(500ms);
```
Comment (wkaluza) [inline libs/core/tests/unit/temporary_blacklist_tests.cpp]: We have an adjustable clock to simulate passage of time in tests. Calls `sleep` in test code bloat execution time unnecessarily.

**Review by wkaluza — CHANGES_REQUESTED** [review]

File: libs/ledger/src/protocols/main_chain_rpc_service.cpp [hunk 2]
```diff
@@ -240,11 +240,18 @@ MainChainRpcService::Address MainChainRpcService::GetRandomTrustedPeer() const
...
+    do
+    {
+      // generate a random peer index
+      std::size_t const index = rng() % direct_peers.size();
 
-    // select the address
-    address = direct_peers[index];
+      // select the address
+      address = direct_peers[index];
+    } while (alien_peers_.IsBlacklisted(address));
```
Comment (wkaluza) [inline libs/ledger/src/protocols/main_chain_rpc_service.cpp:254]: use `std::find_if_not`

**Review by wkaluza — CHANGES_REQUESTED** [review]

File: libs/ledger/src/protocols/main_chain_rpc_service.cpp (hunk 2 above)
Comment (denis-trapeznikov) [reply libs/ledger/src/protocols/main_chain_rpc_service.cpp:254]: > use std::find_if_not

Hmm, @wkaluza, do you suggest employing a custom randomized `ExecutionPolicy`? One problem is, std algorithms accepting policies were only standardised in 17 (and even then most STL implementations did not support them). Even then, that would be an overcomplication of random choice, IMO.

File: libs/core/tests/unit/temporary_blacklist_tests.cpp (hunk 1 above)
Comment (denis-trapeznikov) [reply libs/core/tests/unit/temporary_blacklist_tests.cpp]: > We have an adjustable clock to simulate passage of time in tests.

Where exactly?

Comment (wkaluza) [reply libs/core/tests/unit/temporary_blacklist_tests.cpp]: `adjustable_clock.hpp`

File: libs/ledger/src/protocols/main_chain_rpc_service.cpp (hunk 2 above)
Comment (wkaluza) [reply libs/ledger/src/protocols/main_chain_rpc_service.cpp:254]: You're right, the randomisation would be tricky. Let's leave it as it is, then.

**Review by n-hutton — CHANGES_REQUESTED** [review]
The tests for this segfault

**Pull request closed without merging**